src/library/itunes/itunesmacosimporter.mm

```diff
@@ -81,10 +81,21 @@ void importMediaItems(NSArray<ITLibMediaItem*>* items) {
     }
 
     void appendPlaylistTree(TreeItem& item, int playlistId = -1) {
-        for (auto it = m_playlistChildsByDbId.find(playlistId);
-                it != m_playlistChildsByDbId.end() && it.key() == playlistId;
-                ++it) {
-            int childId = it.value();
+        // Find bounds of the child ID range in the multimap
+        auto childIdsBegin = m_playlistChildsByDbId.constFind(playlistId);
+        auto childIdsEnd = childIdsBegin;
+        while (childIdsEnd != m_playlistChildsByDbId.constEnd() &&
+                childIdsEnd.key() == playlistId) {
+            ++childIdsEnd;
+        }
+
+        // QMultiMap stores items with same key in reverse insertion order,
+        // therefore iterate in reverse to get the insertion order again.
+        // See the docs: https://doc.qt.io/qt-5/qmultimap.html#details
+        auto it = childIdsEnd;
+        while (it != childIdsBegin) {
+            --it;
+            int childId = *it;
```

Comment:
wouldn't this be easier by simply using a `std::reverse_iterator`? Then we don't have to linearly search for the element before the end and then do the looping in reverse. Also, is the `it.key() == playlistId` really necessary? I'd assume that an iterator over a multimap doesn't iterate over unrelated keys. 

```suggestion
        std::for_each(std::make_reverse_iterator(m_playlistChildsByDbId.constEnd()),
            std::make_reverse_iterator(m_playlistChildsByDbId.constFind(playlistId)),
            [](int childId) {....};
```
(this is just a rough outline, I don't know if that would work, I'd highly appreciate if you could experiment. Using `std::ranges::views::reverse` could also be worth looking at.)

Reply:
> Also, is the it.key() == playlistId really necessary?

I'm afraid it is, I've based this on an example [from the docs](https://doc.qt.io/qt-5/qmultimap.html#details) (see the section on STL-style iteration). The reverse iterator sounds interesting, I'll have to look into that (though I'm not sure if it will suffice since the find method only gives us the start of the section, we'd have to seek to the end of the section anyway).

Reply:
> Then we don't have to linearly search for the element before the end and then do the looping in reverse.

Since we have to linearly scan over all the elements anyway, I think traversing them twice shouldn't be an issue.

Reply:
`std::ranges::views::reverse` does look pretty interesting, but was introduced in C++20, can we already use that?

Reply:
Unfortunately this relies on undefined behaviour, because constFind() may find one element in the equal range not necessarily the first.

Reply:
Are you sure? I'm basing this off the documented example

```cpp
QMultiMap<QString, int>::iterator i = map.find("plenty");
while (i != map.end() && i.key() == "plenty") {
    cout << i.value() << Qt::endl;
    ++i;
}
```

If `constFind` would find an arbitrary element instead of the first, it would be nearly useless (or much less intuitive, when would that useful? The user would always have to scan backwards too ib that case)

Reply:
> Are you sure? 

No, I am not sure.. 

It obviously works for Qt5 because QMaps in Qt5 is using a red-black three. And according to the link they seems to guarantee the behaviour for the std::multimap implementations as well. 

The issue is just, that there is no such guarantee. See: https://cplusplus.com/reference/map/multimap/find/

> If constFind would find an arbitrary element instead of the first, it would be nearly useless (or much less intuitive, when would that useful? The user would always have to scan backwards too in that case)

Maybe that is the reason they suggest to use [multimap::equal_range](https://cplusplus.com/multimap::equal_range), where the behaviour is guaranteed. 

But I am not an expert, so I will not stress the issue any further.